### PR TITLE
Add features to friesian/feature/table.py

### PR DESCRIPTION
**Pull request by jiaxinying**
1.In friesian/table.py, add function "get_col_names" which is able to return a list of column names
2.In friesian/table.py, add function "add_constant_values" which is able to add a constant value to all specified columns
3.In friesian/table.py, add function "sample" which is able to sample a subset from the table
4. Write corresponding testing code in test_table.py

### CODE REVIEW

File: pyzoo/zoo/friesian/feature/table.py [hunk 1]
```diff
@@ -291,6 +292,44 @@ def show(self, n=20, truncate=True):
                align cells right.
         """
         self.df.show(n, truncate)
+    
+    def add_constant_values(self, columns, value = 1):
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:296]: just `add`

File: pyzoo/zoo/friesian/feature/table.py [hunk 2]
```diff
@@ -291,6 +292,44 @@ def show(self, n=20, truncate=True):
...
+        """
+        Increase all of values of a column or a list of columns by a constant value
+
+        :param columns: str or list of str, the target columns to be increased
+        :param value: numeric, the constant value to bed added
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:301]: bed -> be

File: pyzoo/zoo/friesian/feature/table.py [hunk 3]
```diff
@@ -291,6 +292,44 @@ def show(self, n=20, truncate=True):
                align cells right.
         """
         self.df.show(n, truncate)
+    
+    def add_constant_values(self, columns, value = 1):
+        """
+        Increase all of values of a column or a list of columns by a constant value
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:298]: comma at the end of each doc line...

File: pyzoo/zoo/friesian/feature/table.py (hunk 2 above)
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:301]: What numeric do we support? int/double/float?

File: pyzoo/zoo/friesian/feature/table.py [hunk 4]
```diff
@@ -291,6 +292,44 @@ def show(self, n=20, truncate=True):
...
+        return self._clone(new_df)
+
+    def get_col_names(self):
+        """
+        Get column names of the table
+        :return: A list of strings that specifies column names 
+        """
+        return self.df.schema.names
+
+    def sample(self, fraction, withReplacement=False, seed = random.random()):
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:322]: withReplacement -> replace

File: pyzoo/zoo/friesian/feature/table.py [hunk 5]
```diff
@@ -292,6 +294,48 @@ def show(self, n=20, truncate=True):
...
+        Get column names of the table
+        :return: A list of strings that specifies column names
+        """
+        return self.df.schema.names
+
+    def sample(self, fraction, replace=False, seed=random.random()):
+        """
+        Return a sampled subset of table
+
+        :param withReplacement: bool, identify if sampled items need to be replaced
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:330]: replace

File: pyzoo/zoo/friesian/feature/table.py [hunk 6]
```diff
@@ -292,6 +294,48 @@ def show(self, n=20, truncate=True):
...
+        new_df = self.df
+        for column in columns:
+            if new_df.schema[column].dataType not in [IntegerType(), ShortType(),
+                                                      LongType(), FloatType(),
+                                                      DecimalType(), DoubleType()]:
+                raise ValueError("column type should be numeric")
+            new_df = new_df.withColumn(column, pyspark_col(column) + value)
+        return self._clone(new_df)
+
+    def get_col_names(self):
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:319]: rename to `columns` ? or make this a property?

File: pyzoo/zoo/friesian/feature/table.py [hunk 7]
```diff
@@ -292,6 +294,48 @@ def show(self, n=20, truncate=True):
         """
         self.df.show(n, truncate)
 
+    def add(self, columns, value=1):
+        """
+        Increase all of values of a column or a list of columns by a constant value
+
+        :param columns: str or list of str, the target columns to be increased
+        :param value: numeric (int/float/double/short/long), the constant value to be added
+        return: A new table that update values of specified columns by a constant value
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:303]: :return   add an empty line before this

File: pyzoo/zoo/friesian/feature/table.py (hunk 1 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py]: fixed

File: pyzoo/zoo/friesian/feature/table.py (hunk 2 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py]: fixed

File: pyzoo/zoo/friesian/feature/table.py (hunk 3 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py]: fixed

File: pyzoo/zoo/friesian/feature/table.py (hunk 2 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py]: support int/double/float/short/long

File: pyzoo/zoo/friesian/feature/table.py (hunk 4 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py]: fixed

File: pyzoo/zoo/friesian/feature/table.py (hunk 7 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py]: fixed

File: pyzoo/zoo/friesian/feature/table.py (hunk 5 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py]: fixed

File: pyzoo/zoo/friesian/feature/table.py (hunk 6 above)
Comment (jiaxinying) [reply pyzoo/zoo/friesian/feature/table.py:320]: This line is to update the column value by the original value in column plus specified value

File: pyzoo/zoo/friesian/feature/table.py [hunk 8]
```diff
@@ -292,6 +294,48 @@ def show(self, n=20, truncate=True):
...
+        return self._clone(new_df)
+
+    def get_col_names(self):
+        """
+        Get column names of the table
+        :return: A list of strings that specifies column names
+        """
+        return self.df.schema.names
+
+    def sample(self, fraction, replace=False, seed=random.random()):
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:328]: seed default to be None?

File: pyzoo/zoo/friesian/feature/table.py [hunk 9]
```diff
@@ -34,6 +36,7 @@
 class Table:
     def __init__(self, df):
         self.df = df
+        self.__column_names = self.df.schema.names
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:39]: _column_names?

File: pyzoo/zoo/friesian/feature/table.py [hunk 10]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
+        for column in columns:
+            if new_df.schema[column].dataType not in [IntegerType(), ShortType(),
+                                                      LongType(), FloatType(),
+                                                      DecimalType(), DoubleType()]:
+                raise ValueError("column type should be numeric")
+            new_df = new_df.withColumn(column, pyspark_col(column) + lit(value))
+        return self._clone(new_df)
+
+    @property
+    def col_names(self):
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:322]: col_names -> columns

File: pyzoo/zoo/friesian/feature/table.py [hunk 11]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
         """
         self.df.show(n, truncate)
 
+    def add(self, columns, value=1):
+        """
+        Increase all of values of a column or a list of columns by a constant value.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:300]: Increase all of values of the target numeric column(s) by a constant value.

File: pyzoo/zoo/friesian/feature/table.py [hunk 12]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
+        if not isinstance(columns, list):
+            columns = [columns]
+        check_col_exists(self.df, columns)
+        new_df = self.df
+        for column in columns:
+            if new_df.schema[column].dataType not in [IntegerType(), ShortType(),
+                                                      LongType(), FloatType(),
+                                                      DecimalType(), DoubleType()]:
+                raise ValueError("column type should be numeric")
+            new_df = new_df.withColumn(column, pyspark_col(column) + lit(value))
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:318]: will it work if value <0?

File: pyzoo/zoo/friesian/feature/table.py [hunk 13]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
         self.df.show(n, truncate)
 
+    def add(self, columns, value=1):
+        """
+        Increase all of values of a column or a list of columns by a constant value.
+
+        :param columns: str or list of str, the target columns to be increased.
+        :param value: numeric (int/float/double/short/long), the constant value to be added.
+
+        return: A new table that update values of specified columns by a constant value.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:305]: table -> Table
A new Table with update numeric values on specified columns.

File: pyzoo/zoo/friesian/feature/table.py [hunk 14]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
+            raise ValueError("Columns should be str or list of str, but got None")
+        if not isinstance(columns, list):
+            columns = [columns]
+        check_col_exists(self.df, columns)
+        new_df = self.df
+        for column in columns:
+            if new_df.schema[column].dataType not in [IntegerType(), ShortType(),
+                                                      LongType(), FloatType(),
+                                                      DecimalType(), DoubleType()]:
+                raise ValueError("column type should be numeric")
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:317]: enrich this error message, for example, print which column in the message.

File: pyzoo/zoo/friesian/feature/table.py [hunk 15]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
+
+    def sample(self, fraction, replace=False, seed=None):
+        """
+        Return a sampled subset of table.
+
+        :param replace: bool, identify if sampled items need to be replaced.
+        during the sample process
+        :param fraction: float, fraction of rows to generate.
+        :param seed: seed for sampling.
+        """
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:338]: missing return

File: pyzoo/zoo/friesian/feature/table.py [hunk 16]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
+        return self.__column_names
+
+    def sample(self, fraction, replace=False, seed=None):
+        """
+        Return a sampled subset of table.
+
+        :param replace: bool, identify if sampled items need to be replaced.
+        during the sample process
+        :param fraction: float, fraction of rows to generate.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:336]: put fraction before replace.

File: pyzoo/zoo/friesian/feature/table.py [hunk 17]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
+        :return: A list of strings that specifies column names.
+        """
+        return self.__column_names
+
+    def sample(self, fraction, replace=False, seed=None):
+        """
+        Return a sampled subset of table.
+
+        :param replace: bool, identify if sampled items need to be replaced.
+        during the sample process
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:335]: missing . at the end

File: pyzoo/zoo/friesian/feature/table.py [hunk 18]
```diff
@@ -292,6 +295,51 @@ def show(self, n=20, truncate=True):
...
+        :param fraction: float, fraction of rows to generate.
+        :param seed: seed for sampling.
+        """
+        if fraction < 0 or fraction > 1:
+            raise ValueError("fraction should in the range of [0,1]")
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:340]: should be within the range of [0, 1]
space before ,1

File: pyzoo/zoo/friesian/feature/table.py (hunk 17 above)
Comment (hkvision) [reply pyzoo/zoo/friesian/feature/table.py:335]: Allow or disallow sampling of the same row more than once.

File: pyzoo/zoo/friesian/feature/table.py [hunk 19]
```diff
@@ -292,6 +295,53 @@ def show(self, n=20, truncate=True):
...
+
+    def sample(self, fraction, replace=False, seed=None):
+        """
+        Return a sampled subset of Table.
+
+        :param fraction: float, fraction of rows to generate.
+        :param replace: allow or disallow sampling of the same row more than once.
+        :param seed: seed for sampling.
+
+        return: a sampled subset of Table
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:339]: :return: A new Table with sampled rows.

File: pyzoo/zoo/friesian/feature/table.py [hunk 20]
```diff
@@ -292,6 +295,53 @@ def show(self, n=20, truncate=True):
...
+        """
+        if fraction < 0 or fraction > 1:
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:341]: will spark check this and raise error?

File: pyzoo/test/zoo/friesian/feature/test_table.py [hunk 21]
```diff
@@ -524,5 +525,46 @@ def test_encode_string_from_dict(self):
...
+        columns = ["age", "height"]
+        new_tbl = tbl.add(columns, 1.5)
+        new_list = new_tbl.df.take(3)
+        assert len(new_list) == 3, "new_tbl should have 3 rows"
+        assert new_list[0]['age'] == 15.5, "the age of jack should increase 1"
+        assert new_list[0]['height'] == 10, "the height of jack should increase 1"
+        assert new_list[1]['age'] == 26.5, "the age of alice should increase 1"
+        assert new_list[1]['height'] == 11.1, "the height of alice should increase 1"
+        assert new_list[2]['age'] == 24.5, "the age of rose should increase 1"
+        assert new_list[2]['height'] == 11.5, "the height of rose should increase 1"
```
Comment (hkvision) [inline pyzoo/test/zoo/friesian/feature/test_table.py:555]: should be 1.5 not 1

File: pyzoo/test/zoo/friesian/feature/test_table.py [hunk 22]
```diff
@@ -524,5 +525,46 @@ def test_encode_string_from_dict(self):
...
+        data = [("jack", "123", 14, 8.5),
+                ("alice", "34", 25, 9.6),
+                ("rose", "25344", 23, 10.0)]
+        schema = StructType([StructField("name", StringType(), True),
+                             StructField("num", StringType(), True),
+                             StructField("age", IntegerType(), True),
+                             StructField("height", DoubleType(), True)])
+        tbl = FeatureTable(spark.createDataFrame(data, schema))
+        columns = ["age", "height"]
+        new_tbl = tbl.add(columns, 1.5)
```
Comment (hkvision) [inline pyzoo/test/zoo/friesian/feature/test_table.py:547]: add a unit test to add integers/ and negative values...

File: pyzoo/zoo/friesian/feature/table.py [hunk 23]
```diff
@@ -293,6 +295,53 @@ def show(self, n=20, truncate=True):
...
+
+        :return: A list of strings that specifies column names.
+        """
+        return self.__column_names
+
+    def sample(self, fraction, replace=False, seed=None):
+        """
+        Return a sampled subset of Table.
+
+        :param fraction: float, fraction of rows to generate, fraction should be
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:335]: , should be...

File: pyzoo/zoo/friesian/feature/table.py [hunk 24]
```diff
@@ -293,6 +295,53 @@ def show(self, n=20, truncate=True):
...
         self.df.show(n, truncate)
 
+    def add(self, columns, value=1):
+        """
+        Increase all of values of the target numeric column(s) by a constant value.
+
+        :param columns: str or list of str, the target columns to be increased.
+        :param value: numeric (int/float/double/short/long), the constant value to be added.
+
+        return: A new Table with update numeric values on specified columns.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:305]: updated

File: pyzoo/zoo/friesian/feature/table.py [hunk 25]
```diff
@@ -293,6 +295,52 @@ def show(self, n=20, truncate=True):
...
+    def sample(self, fraction, replace=False, seed=None):
+        """
+        Return a sampled subset of Table.
+
+        :param fraction: float, fraction of rows to generate, should be within the 
+        range [0, 1].
+        :param replace: allow or disallow sampling of the same row more than once.
+        :param seed: seed for sampling.
+
+        return: A new Table with sampled rows.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:340]: :return

File: pyzoo/zoo/friesian/feature/table.py (hunk 24 above)
Comment (hkvision) [reply pyzoo/zoo/friesian/feature/table.py:305]: :return

File: pyzoo/zoo/friesian/feature/table.py [hunk 26]
```diff
@@ -293,6 +295,52 @@ def show(self, n=20, truncate=True):
...
+                    for column {column}".format(new_df.schema[column].dataType, column))
+            new_df = new_df.withColumn(column, pyspark_col(column) + lit(value))
+        return self._clone(new_df)
+
+    @property
+    def columns(self):
+        """
+        Get column names of the Table.
+
+        :return: A list of strings that specifies column names.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:327]: specifies -> specify

File: pyzoo/zoo/friesian/feature/table.py [hunk 27]
```diff
@@ -293,6 +295,52 @@ def show(self, n=20, truncate=True):
...
+            raise ValueError("Columns should be str or list of str, but got None")
+        if not isinstance(columns, list):
+            columns = [columns]
+        check_col_exists(self.df, columns)
+        new_df = self.df
+        for column in columns:
+            if new_df.schema[column].dataType not in [IntegerType(), ShortType(),
+                                                      LongType(), FloatType(),
+                                                      DecimalType(), DoubleType()]:
+                raise ValueError("Column type should be numeric, but have type {type} \
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:317]: is this syntax correct?

File: pyzoo/test/zoo/friesian/feature/test_table.py [hunk 28]
```diff
@@ -554,6 +605,5 @@ def test_group_by(self):
             .count() == feature_tbl.df.filter("col_4 == 'b' and col_5 == 'dd'").count(), \
             "first of col_1 should be 0 for all col_4 = 'b' and col_5 = 'dd' in groupby_tbl4"
 
-
```
Comment (hkvision) [inline pyzoo/test/zoo/friesian/feature/test_table.py:557]: revert this deletion?

**Review by hkvision — APPROVED** [review]

**Comment by jiaxinying** [comment]
http://10.239.176.111:18888/job/ZOO-PR-Validation/5867/
